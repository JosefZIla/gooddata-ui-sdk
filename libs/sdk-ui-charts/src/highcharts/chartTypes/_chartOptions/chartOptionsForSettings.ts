// (C) 2007-2022 GoodData Corporation

import { ISettings } from "@gooddata/sdk-model";
import { IChartConfig } from "../../../interfaces";

/**
 * @internal
 */
export function updateConfigWithSettings(config: IChartConfig, settings: ISettings): IChartConfig {
    let updatedConfig = config;
    if (settings) {
        if (settings["disableKpiDashboardHeadlineUnderline"] === true) {
            updatedConfig = { ...updatedConfig, disableDrillUnderline: true };
        }

        if (settings["enableKDWidgetCustomHeight"] === true) {
            updatedConfig = { ...updatedConfig, enableCompactSize: true };
        }

        if (updatedConfig === undefined || updatedConfig.enableJoinedAttributeAxisName === undefined) {
            updatedConfig = {
                ...updatedConfig,
                enableJoinedAttributeAxisName: settings["enableAxisNameViewByTwoAttributes"],
            };
        }

        if (settings.enableChartsSorting) {
            updatedConfig = { ...updatedConfig, enableChartSorting: true };
        }
    }

    return updatedConfig;
}
