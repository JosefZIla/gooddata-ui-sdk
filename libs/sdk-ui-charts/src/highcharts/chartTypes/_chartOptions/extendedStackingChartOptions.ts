// (C) 2007-2022 GoodData Corporation
import { IUnwrappedAttributeHeadersWithItems } from "../../typings/mess";
import { IResultAttributeHeader } from "@gooddata/sdk-model";

type NameAndCategories = {
    name: string;
    categories: string[];
};

/**
 * Transform
 *      viewByParentAttribute: [P1, P1, P2, P2, P3],
 *      viewByAttribute: [C1, C2, C1, C2, C2]
 * to
 * ```
 *     [{
 *         name: P1,
 *         categories: [C1, C2]
 *      }, {
 *         name: P2,
 *         categories: [C1, C2]
 *      }, {
 *         name: P3,
 *         categories: [C2]
 *      }]
 * ```
 */
export function getCategoriesForTwoAttributes(
    viewByAttribute: IUnwrappedAttributeHeadersWithItems,
    viewByParentAttribute: IUnwrappedAttributeHeadersWithItems,
): NameAndCategories[] {
    const keys: string[] = [];
    const { items: children } = viewByAttribute;
    const { items: parent } = viewByParentAttribute;

    const combinedResult = parent.reduce(
        (
            result: { [property: string]: NameAndCategories },
            parentAttr: IResultAttributeHeader,
            index: number,
        ) => {
            const uri = parentAttr?.attributeHeaderItem?.uri ?? "";
            const name = parentAttr?.attributeHeaderItem?.name ?? "";
            const value = children[index]?.attributeHeaderItem?.name ?? "";

            const existingEntry = result[uri];
            const childCategories = existingEntry?.categories ?? [];

            if (!childCategories.length) {
                keys.push(uri);
            }

            return {
                ...result,
                [uri]: {
                    name,
                    categories: [...childCategories, value], // append value
                },
            };
        },
        {},
    );

    return keys.map((key: string) => {
        const { name, categories } = combinedResult[key];

        return {
            name,
            categories,
        };
    });
}
