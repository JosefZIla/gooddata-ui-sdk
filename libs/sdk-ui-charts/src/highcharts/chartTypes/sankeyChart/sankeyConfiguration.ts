// (C) 2023 GoodData Corporation
import cloneDeep from "lodash/cloneDeep";
import { getCommonResponsiveConfig } from "../_chartCreators/responsive";
import { HighchartsOptions } from "../../lib";

const SANKEY_TEMPLATE = {
    chart: {
        type: "sankey",
    },
    plotOptions: {
        sankey: {
            showInLegend: true,
            dataLabels: {
                enabled: true,
                padding: 12,
                align: "left",
            },
            nodeWidth: 5,
        },
    },
    legend: {
        enabled: false,
    },
};

export function getSankeyConfiguration(): HighchartsOptions {
    const configuration = cloneDeep(SANKEY_TEMPLATE);

    return {
        ...configuration,
        responsive: getCommonResponsiveConfig(),
    };
}
