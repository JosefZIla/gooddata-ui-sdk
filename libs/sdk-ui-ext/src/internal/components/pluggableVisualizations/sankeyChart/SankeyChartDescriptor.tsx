// (C) 2022-2023 GoodData Corporation
import { BucketNames } from "@gooddata/sdk-ui";
import { ISankeyChartProps } from "@gooddata/sdk-ui-charts";
import {
    IVisualizationDescriptor,
    IVisualizationMeta,
    PluggableVisualizationFactory,
} from "../../../interfaces/VisualizationDescriptor";
import { PluggableSankeyChart } from "./PluggableSankeyChart";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
import {
    executionConfigInsightConversion,
    filtersInsightConversion,
    getInsightToPropsConverter,
    getReactEmbeddingCodeGenerator,
    localeInsightConversion,
    singleAttributeBucketConversion,
    singleAttributeOrMeasureBucketConversion,
    sortsInsightConversion,
} from "../../../utils/embeddingCodeGenerator";
import { chartAdditionalFactories, chartConfigInsightConversion } from "../chartCodeGenUtils";

export class SankeyChartDescriptor extends BaseChartDescriptor implements IVisualizationDescriptor {
    public getFactory(): PluggableVisualizationFactory {
        return (params) => new PluggableSankeyChart(params);
    }

    public getEmbeddingCode = getReactEmbeddingCodeGenerator({
        component: {
            importType: "named",
            name: "SankeyChart",
            package: "@gooddata/sdk-ui-charts",
        },
        insightToProps: getInsightToPropsConverter<ISankeyChartProps>({
            measure: singleAttributeOrMeasureBucketConversion("measure", BucketNames.MEASURES),
            attributeFrom: singleAttributeBucketConversion("attributeFrom", BucketNames.ATTRIBUTE_FROM),
            attributeTo: singleAttributeBucketConversion("attributeTo", BucketNames.ATTRIBUTE_FROM),
            filters: filtersInsightConversion("filters"),
            sortBy: sortsInsightConversion("sortBy"),
            config: chartConfigInsightConversion("config"),
            locale: localeInsightConversion("locale"),
            execConfig: executionConfigInsightConversion("execConfig"),
        }),
        additionalFactories: chartAdditionalFactories(),
    });

    public getMeta(): IVisualizationMeta {
        return {
            supportsExport: true,
            supportsZooming: false,
        };
    }
}
