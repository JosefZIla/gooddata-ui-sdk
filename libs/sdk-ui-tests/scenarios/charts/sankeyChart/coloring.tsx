// (C) 2007-2019 GoodData Corporation
import { ISankeyChartProps, SankeyChart } from "@gooddata/sdk-ui-charts";
import { scenariosFor } from "../../../src";
import { BlackColor, CustomColorPalette, RedColor } from "../../_infra/colors";
import { AttributeElements } from "../../_infra/predicates";
import { coloringCustomizer } from "../_infra/coloringVariants";
import { replaceMappingPredicates } from "../_infra/insightConverters";
import { ReferenceData } from "@gooddata/reference-workspace";
import { ScenarioGroupNames } from "../_infra/groupNames";
import { SankeyChartWithMeasureAttributeFromAndTo } from "./base";

const colorsAndPalette = scenariosFor<ISankeyChartProps>("SankeyChart", SankeyChart)
    .withGroupNames(...ScenarioGroupNames.Coloring)
    .withVisualTestConfig({ groupUnder: "coloring" })
    .withDefaultTags("vis-config-only", "mock-no-scenario-meta")
    .addScenarios("coloring", SankeyChartWithMeasureAttributeFromAndTo, coloringCustomizer);

const colorAssignment = scenariosFor<ISankeyChartProps>("SankeyChart", SankeyChart)
    .withGroupNames(...ScenarioGroupNames.Coloring)
    .withDefaultTags("vis-config-only", "mock-no-scenario-meta")
    .addScenario(
        "assign color to nodes",
        {
            ...SankeyChartWithMeasureAttributeFromAndTo,
            config: {
                colorPalette: CustomColorPalette,
                colorMapping: [
                    {
                        predicate: AttributeElements.Product.WonderKid,
                        color: BlackColor,
                    },
                    {
                        predicate: AttributeElements.Product.Explorer,
                        color: RedColor,
                    },
                ],
            },
        },
        (m) =>
            m.withInsightConverter(
                replaceMappingPredicates(
                    ReferenceData.ProductName.WonderKid.uri,
                    ReferenceData.ProductName.Explorer.uri,
                ),
            ),
    );

export default [colorsAndPalette, colorAssignment];
