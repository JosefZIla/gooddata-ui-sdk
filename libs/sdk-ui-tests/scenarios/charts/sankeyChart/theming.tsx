// (C) 2023 GoodData Corporation
import { ISankeyChartProps, SankeyChart } from "@gooddata/sdk-ui-charts";
import { scenariosFor } from "../../../src";
import { ScenarioGroupNames } from "../_infra/groupNames";
import { SankeyChartWithMeasureAttributeFromAndTo } from "./base";

export default scenariosFor<ISankeyChartProps>("SankeyChart", SankeyChart)
    .withGroupNames(...ScenarioGroupNames.Theming)
    .withDefaultTestTypes("visual")
    .withDefaultTags("themed")
    .addScenario("themed", SankeyChartWithMeasureAttributeFromAndTo)
    .addScenario("font", SankeyChartWithMeasureAttributeFromAndTo, (m) => m.withTags("themed", "font"));
