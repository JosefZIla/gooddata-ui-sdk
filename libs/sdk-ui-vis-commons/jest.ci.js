// (C) 2019 GoodData Corporation
/* eslint-disable @typescript-eslint/no-var-requires */
const ciBase = require("../../common/config/jest/jest.config.ci.base.js");

module.exports = {
    ...ciBase,
    testEnvironment: "jsdom",
    setupFilesAfterEnv: ["<rootDir>/jest.setup.ts"],
};
