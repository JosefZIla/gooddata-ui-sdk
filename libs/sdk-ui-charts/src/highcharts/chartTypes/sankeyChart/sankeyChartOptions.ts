// (C) 2023 GoodData Corporation
import { ISeriesItem } from "../../typings/unsafe";
import { DataViewFacade, getMappingHeaderFormattedName, IColorAssignment } from "@gooddata/sdk-ui";
import { IUnwrappedAttributeHeadersWithItems } from "../../typings/mess";
import { getColorByGuid, getRgbStringFromRGB, valueWithEmptyHandling } from "@gooddata/sdk-ui-vis-commons";
import { IColorPalette, isRgbColor } from "@gooddata/sdk-model";
import compact from "lodash/compact";

const KEYS = ["from", "to", "weight", "name"];
const DEFAULT_ATTRIBUTE_HEADER = "";
const DEFAULT_WEIGHT = 1;

const buildNodes = (
    colorAssignments: IColorAssignment[],
    colorPalette: IColorPalette,
    emptyHeaderTitle: string,
) => {
    return colorAssignments.map((colorAssignment) => ({
        id: valueWithEmptyHandling(
            getMappingHeaderFormattedName(colorAssignment.headerItem),
            emptyHeaderTitle,
        ),
        color: isRgbColor(colorAssignment.color)
            ? getRgbStringFromRGB(colorAssignment.color.value)
            : getRgbStringFromRGB(getColorByGuid(colorPalette, colorAssignment.color.value, 0)),
    }));
};

const buildEmptyData = (dv: DataViewFacade) => {
    return [
        {
            from: DEFAULT_ATTRIBUTE_HEADER,
            to: DEFAULT_ATTRIBUTE_HEADER,
            weight: Number((dv.rawData().dataAt(0) as Array<number>)[0]) || DEFAULT_WEIGHT,
            name: DEFAULT_ATTRIBUTE_HEADER,
        },
    ];
};

const buildData = (
    dv: DataViewFacade,
    attributeHeaders: IUnwrappedAttributeHeadersWithItems[],
    emptyHeaderTitle: string,
) => {
    const [from, to] = compact(attributeHeaders);
    const isExistingAttributeHeader = from || to;

    if (!isExistingAttributeHeader) {
        return buildEmptyData(dv);
    }

    return (dv.rawData().dataAt(0) as Array<number>).map((it, index) => {
        const fromValue =
            from &&
            valueWithEmptyHandling(getMappingHeaderFormattedName(from.items[index]), emptyHeaderTitle);

        const toValue =
            to && valueWithEmptyHandling(getMappingHeaderFormattedName(to.items[index]), emptyHeaderTitle);

        return {
            from: fromValue,
            to: toValue,
            weight: Number(it),
            name: fromValue,
        };
    });
};

export const buildSankeyChartSeries = (
    dv: DataViewFacade,
    attributeHeaders: IUnwrappedAttributeHeadersWithItems[],
    colorAssignments: IColorAssignment[],
    colorPalette: IColorPalette,
    emptyHeaderTitle: string,
): ISeriesItem[] => {
    return [
        {
            keys: KEYS,
            data: buildData(dv, attributeHeaders, emptyHeaderTitle),
            nodes: buildNodes(colorAssignments, colorPalette, emptyHeaderTitle),
        },
    ];
};
