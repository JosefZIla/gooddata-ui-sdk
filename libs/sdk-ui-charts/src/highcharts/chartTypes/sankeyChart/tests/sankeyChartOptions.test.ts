// (C) 2023 GoodData Corporation
import { recordedDataFacade } from "../../../../../__mocks__/recordings";
import { getMVSForViewByTwoAttributes } from "../../_util/test/helper";
import { IColorStrategy } from "@gooddata/sdk-ui-vis-commons";
import { ColorFactory } from "../../_chartOptions/colorFactory";
import { CUSTOM_COLOR_PALETTE } from "../../_util/test/colorPalette.fixture";
import { buildSankeyChartSeries } from "../sankeyChartOptions";
import { CHART_TYPE, COLOR_MAPPINGS, EMPTY_HEADER_TITLE_VALUE, RECORDS } from "./sankeyChart.fixture";

describe("SankeyChart buildSankeyChartSeries", () => {
    it.each(RECORDS)("should return series correctly in cases %s ", (_title, record) => {
        const dv = recordedDataFacade(record);
        const { viewByAttribute, viewByParentAttribute } = getMVSForViewByTwoAttributes(dv);
        const colorStrategy: IColorStrategy = ColorFactory.getColorStrategy(
            CUSTOM_COLOR_PALETTE,
            undefined,
            viewByAttribute,
            viewByParentAttribute,
            null,
            dv,
            CHART_TYPE,
        );
        const series = buildSankeyChartSeries(
            dv,
            [viewByParentAttribute, viewByAttribute],
            colorStrategy.getColorAssignment(),
            CUSTOM_COLOR_PALETTE,
            EMPTY_HEADER_TITLE_VALUE,
        );
        expect(series).toMatchSnapshot();
    });

    it.each(RECORDS)(
        "should return series correctly with custom color mappings in cases %s ",
        (_title, record) => {
            const dv = recordedDataFacade(record);
            const { viewByAttribute, viewByParentAttribute } = getMVSForViewByTwoAttributes(dv);

            const colorStrategy: IColorStrategy = ColorFactory.getColorStrategy(
                CUSTOM_COLOR_PALETTE,
                COLOR_MAPPINGS,
                viewByAttribute,
                viewByParentAttribute,
                null,
                dv,
                CHART_TYPE,
            );
            const series = buildSankeyChartSeries(
                dv,
                [viewByParentAttribute, viewByAttribute],
                colorStrategy.getColorAssignment(),
                CUSTOM_COLOR_PALETTE,
                EMPTY_HEADER_TITLE_VALUE,
            );
            expect(series).toMatchSnapshot();
        },
    );
});
