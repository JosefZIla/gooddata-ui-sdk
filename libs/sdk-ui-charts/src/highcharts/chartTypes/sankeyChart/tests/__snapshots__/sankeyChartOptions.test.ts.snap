// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SankeyChart buildSankeyChartSeries should return series correctly in cases 1 measure and 0 attribute  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "",
        "name": "",
        "to": "",
        "weight": 116625456.54,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly in cases 1 measure and 1 attribute_from  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": undefined,
        "weight": 27222899.64,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": undefined,
        "weight": 22946895.47,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": undefined,
        "weight": 38596194.86,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": undefined,
        "weight": 8042031.92,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": undefined,
        "weight": 9525857.91,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": undefined,
        "weight": 10291576.74,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "CompuSci",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "Educationly",
      },
      Object {
        "color": "rgb(243,217,177)",
        "id": "Explorer",
      },
      Object {
        "color": "rgb(194,153,121)",
        "id": "Grammar Plus",
      },
      Object {
        "color": "rgb(162,37,34)",
        "id": "PhoenixSoft",
      },
      Object {
        "color": "rgb(195,49,73)",
        "id": "WonderKid",
      },
    ],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly in cases 1 measure and 1 attribute_to  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "East Coast",
        "name": "East Coast",
        "to": undefined,
        "weight": 28017096.42,
      },
      Object {
        "from": "West Coast",
        "name": "West Coast",
        "to": undefined,
        "weight": 88608360.12,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "East Coast",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "West Coast",
      },
    ],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly in cases 1 measure and 2 attributes  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": "East Coast",
        "weight": 5099309.54,
      },
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": "West Coast",
        "weight": 22123590.1,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": "East Coast",
        "weight": 6614222.47,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": "West Coast",
        "weight": 16332673,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": "East Coast",
        "weight": 8269682.68,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": "West Coast",
        "weight": 30326512.18,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": "East Coast",
        "weight": 2761931.26,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": "West Coast",
        "weight": 5280100.66,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": "East Coast",
        "weight": 2664832.4,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": "West Coast",
        "weight": 6861025.51,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": "East Coast",
        "weight": 2607118.07,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": "West Coast",
        "weight": 7684458.67,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "CompuSci",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "Educationly",
      },
      Object {
        "color": "rgb(243,217,177)",
        "id": "Explorer",
      },
      Object {
        "color": "rgb(194,153,121)",
        "id": "Grammar Plus",
      },
      Object {
        "color": "rgb(162,37,34)",
        "id": "PhoenixSoft",
      },
      Object {
        "color": "rgb(195,49,73)",
        "id": "WonderKid",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "East Coast",
      },
      Object {
        "color": "rgb(243,217,177)",
        "id": "West Coast",
      },
    ],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly with custom color mappings in cases 1 measure and 0 attribute  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "",
        "name": "",
        "to": "",
        "weight": 116625456.54,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly with custom color mappings in cases 1 measure and 1 attribute_from  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": undefined,
        "weight": 27222899.64,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": undefined,
        "weight": 22946895.47,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": undefined,
        "weight": 38596194.86,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": undefined,
        "weight": 8042031.92,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": undefined,
        "weight": 9525857.91,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": undefined,
        "weight": 10291576.74,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "CompuSci",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "Educationly",
      },
      Object {
        "color": "rgb(255,0,0)",
        "id": "Explorer",
      },
      Object {
        "color": "rgb(194,153,121)",
        "id": "Grammar Plus",
      },
      Object {
        "color": "rgb(162,37,34)",
        "id": "PhoenixSoft",
      },
      Object {
        "color": "rgb(0,0,0)",
        "id": "WonderKid",
      },
    ],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly with custom color mappings in cases 1 measure and 1 attribute_to  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "East Coast",
        "name": "East Coast",
        "to": undefined,
        "weight": 28017096.42,
      },
      Object {
        "from": "West Coast",
        "name": "West Coast",
        "to": undefined,
        "weight": 88608360.12,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "East Coast",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "West Coast",
      },
    ],
  },
]
`;

exports[`SankeyChart buildSankeyChartSeries should return series correctly with custom color mappings in cases 1 measure and 2 attributes  1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": "East Coast",
        "weight": 5099309.54,
      },
      Object {
        "from": "CompuSci",
        "name": "CompuSci",
        "to": "West Coast",
        "weight": 22123590.1,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": "East Coast",
        "weight": 6614222.47,
      },
      Object {
        "from": "Educationly",
        "name": "Educationly",
        "to": "West Coast",
        "weight": 16332673,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": "East Coast",
        "weight": 8269682.68,
      },
      Object {
        "from": "Explorer",
        "name": "Explorer",
        "to": "West Coast",
        "weight": 30326512.18,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": "East Coast",
        "weight": 2761931.26,
      },
      Object {
        "from": "Grammar Plus",
        "name": "Grammar Plus",
        "to": "West Coast",
        "weight": 5280100.66,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": "East Coast",
        "weight": 2664832.4,
      },
      Object {
        "from": "PhoenixSoft",
        "name": "PhoenixSoft",
        "to": "West Coast",
        "weight": 6861025.51,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": "East Coast",
        "weight": 2607118.07,
      },
      Object {
        "from": "WonderKid",
        "name": "WonderKid",
        "to": "West Coast",
        "weight": 7684458.67,
      },
    ],
    "keys": Array [
      "from",
      "to",
      "weight",
      "name",
    ],
    "nodes": Array [
      Object {
        "color": "rgb(195,49,73)",
        "id": "CompuSci",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "Educationly",
      },
      Object {
        "color": "rgb(255,0,0)",
        "id": "Explorer",
      },
      Object {
        "color": "rgb(194,153,121)",
        "id": "Grammar Plus",
      },
      Object {
        "color": "rgb(162,37,34)",
        "id": "PhoenixSoft",
      },
      Object {
        "color": "rgb(0,0,0)",
        "id": "WonderKid",
      },
      Object {
        "color": "rgb(168,194,86)",
        "id": "East Coast",
      },
      Object {
        "color": "rgb(243,217,177)",
        "id": "West Coast",
      },
    ],
  },
]
`;
